src/goto-programs/abstract-interpretation/interval_domain.cpp

```diff
@@ -11,52 +11,79 @@
 #endif
 // Let's start with all templates specializations.
 
+template <size_t Index, class Interval>
+Interval interval_domaint::get_interval_from_variant(const symbol2t &sym) const
+{
+  const auto &it = intervals.find(sym.thename);
+  // TODO: mix floats/integer
+  if (it != intervals.end() && it->second.index() == Index)
```

Comment:
Can the second condition ever fail here? If not, it might be possible to drop the `class Interval` parameter and use `decltype(std::get<Index>(declval<interval>()))` instead. If yes, I'd suggest to add a comment describing why/how it could fail (some mismatch between Index and the type stored in the variant).

Reply:
This might happen due to some wrong logic e.g., calling get<0> on a float interval.

Reply:
std::get\<..\> will catch this (throws [std::bad_variant_access](https://en.cppreference.com/w/cpp/utility/variant/bad_variant_access)). If this guarding against logic errors, the function shouldn't return a default value in that case but abort.

Reply:
Hm, I tried a bit with:
 
 ```cpp
template <size_t Index>
decltype(*(std::get<Index>(std::declval<interval_domaint::interval>())))
interval_domaint::get_interval_from_variant(const symbol2t &sym) const
{
  const auto &it = intervals->find(sym.thename);
  if (it != intervals->end())
  {
    return *std::get<Index>(it->second);
  }
  using IntervalRef = decltype(*std::get<Index>(std::declval<interval_domaint::interval>()));
  return IntervalRef(sym.type);
}
```

But it seems that the bottom bit is giving me a segmentation fault. Mainly because that type is a reference. I might consider just returning a shared ptr though.

Reply:
Nah, that's different than what I intended, I'll give it a try tomorrow.

Reply:
OK, never mind, I don't have a better solution right now. Thanks for addressing the second condition in the if-stmt.